lib/cfa/augeas_parser/writer.rb

```diff
@@ -0,0 +1,155 @@
+        key, path = key_and_path(tree, entry, prefix)
+        process_operation(tree, entry, last_valid_entry_path, key, path)
+        last_valid_entry_path = path if entry[:operation] != :remove
+      end
+    end
+
+  private
+
+    attr_reader :aug
+
+    # returns available number for given collection that can be used to write
+    # @param tree [CFA::AugeasTree] tree where collection is defined
+    # @param key [String] collection name
+    def new_array_number(tree, key)
+      all_elements = tree.data(filtered: false)
+                         .select { |e| e[:key] == key }
+      nums = all_elements.map do |entry|
+        entry[:orig_key] ? entry[:orig_key][/^.*\[(\d+)\]$/, 1].to_i : 0
+      end
+      nums.max ? (nums.max + 1) : 1
+    end
+
+    # construct augeas key and whole path where to write for collection
+    def key_and_path_array(tree, entry, prefix)
+      if entry[:orig_key]
+        key = entry[:orig_key]
+      else
+        strip_key = entry[:key]
+        key = strip_key[0..-2] + new_array_number(tree, strip_key).to_s + "]"
+      end
+      path = prefix + "/" + key
+      [key, path]
+    end
+
+    # construct augeas key and whole path where to write
+    def key_and_path(tree, entry, prefix)
+      if entry[:key].end_with?("[]")
+        key_and_path_array(tree, entry, prefix)
+      else
+        [entry[:key], prefix + "/" + entry[:key]]
+      end
+    end
+
+    # it insert position for given key. Do not set its value.
+    # Its logic is to set it after last valid entry. If it is not defined
+    # then try to place it before first valid entry in tree. If there is
+    # no entry in tree, then do not insert position, which means, that
+    # following setting of value appends it to the end.
+    #
+    # @param last [String] path of last valid entry written to tree
+    # @param key [String] key that need to be inserted
+    # @param tree [CFA::AugeasTree] tree where is key located
+    def insert_entry(last, key, tree)
+      if last
+        report_error { aug.insert(last, key, false) }
+      else
+        e = find_first_valid(tree.data(filtered: false))
+        if e
+          e_path = prefix + "/" + e[:orig_key]
+          report_error { aug.insert(e_path, key, true) }
+        end
+      end
+    end
+
+    # Do modification according to operation defined in AugeasEntry
+    # @param tree [CFA::AugeasTree] tree where entry lives
+    # @param entry [AugeasElement] entry to process
+    # @param last_valid_entry_path [String, nil] path of last valid entry
+    #   written or nil if there is no such entry
+    # @param key [String] valid augeas key for entry. Important for collections
+    #   which have key without its index, but augeas do not allow it.
+    # @param path [String] whole path where to write entry in augeas
+    def process_operation(tree, entry,
+      last_valid_entry_path, key, path)
+      case entry[:operation]
+      when :add, nil # add is default operation
```

Comment:
yes, for backward compatibility. We allow to create own subtree and append it to tree. In this case all new manually crafted values are new and need to be added.